### PR TITLE
Experimental 2020A and 2020B Modes

### PR DESCRIPTION
**Pull request by drowe67**
Experimental 2020A and 2020B modes as discussed in https://github.com/drowe67/codec2/discussions/217.  Also see associated [LPCNet PR](https://github.com/drowe67/LPCNet/pull/42).

Octave simulation (MPP channel at 10dB SNR):
```
octave:68> ofdm_tx("test_2020A.raw", "2020A",60,10,'mpp')
octave:68> ofdm_rx("test_2020A.raw", "2020A")
``` 

C simulation (AWGN channel at 4dB SNR):
```
 ./src/ofdm_mod --in /dev/zero --testframes 10 --mode 2020A | ./src/cohpsk_ch - - -22 --Fs 8000 | ./src/ofdm_demod --mode 2020A --testframes -v1 > /dev/null
```
C simulation with compression, we can drop the noise level for similar BERs with the same peak modulator output:
```
./src/ofdm_mod --in /dev/zero --testframes 30 --mode 2020A --clip | ./src/cohpsk_ch - - -18 --Fs 8000 | ./src/ofdm_demod --mode 2020A --testframes -v1 > /dev/null
```
LPCNet passing through 2020A modem, on a MPP (1 Hz fading, 2ms delay) channel at around 14dB SNR.  Audio is rough but intelligible, also works on MPD channel (2 Hz fading, 4ms delay):
```
cat ~/LPCNet/wav/wia.wav | ~/LPCNet/build_linux/src/lpcnet_enc -s | ./ofdm_mod --mode 2020A -v1 --clip | ./cohpsk_ch - - -26 --Fs 8000 --mpp --fading_dir ../unittest | ./ofdm_demod --mode 2020A --verbose 1 | ~/LPCNet/build_linux/src/lpcnet_dec -s | aplay -f S16_LE -r 16000
```

TODO:
1. Waveform design on spreadsheet :heavy_check_mark: 
1. Octave and C simulations :heavy_check_mark: 
2. Compression for 2020A (and 2020) :heavy_check_mark: 
3. Run with LPCNet on command line :heavy_check_mark: 
4. Compare performance with 2020 at a couple of SNR/channel operating points. Initial results: (i) on MPP channel around 14dB SNR with compression added to both 2020 and 2020A, I prefer 2020.  This makes sense as 2020A has no FEC and runs at a lower Eb/No for a given SNR.  (ii) on MPD channel 14dB SNR 2020 breaks, 2020A hangs on OK.  This is expected, 2020A was designed for fast multipath fading. :heavy_check_mark: 
5. Try [index optimisation](https://github.com/drowe67/LPCNet/pull/42) of VQ tables in LPCNet. Well index optimisation didn't help much, but this work has suggested a [partial FEC protection](https://github.com/drowe67/LPCNet/pull/42#issuecomment-1003471575) design for both 2020 and 2020A.  Although I guess this would break 2020 so we might need a new mode name. :heavy_check_mark: 
1. Build up partial FEC protection design and test on command line: this [seems to work pretty well](https://github.com/drowe67/codec2/pull/274#issuecomment-1011450132), worth trying OTA. :heavy_check_mark: 
2. Draw conclusions and decide on best algorithm to move forward with. Two candidates have emerged, so have changed nomenclature.  Have decided to move forward with (i) 2020A - 2020 waveform but unequal error protection (ii) 2020B - new waveform with faster pilot symbol rate and also employing unequal error correction :heavy_check_mark: 
1. FreeDV API integration of 2020A. :heavy_check_mark: 
1. Prototype of 2020B. :heavy_check_mark: 
1. FreeDV API integration of 2020B. :heavy_check_mark: 
1. Update `README_freedv.md` for new modes.  :heavy_check_mark: 
1. Script based tests of 2020A/2020B using a few samples. Evaluate results, and plan next steps. :heavy_check_mark: 
5. Automated OTA 2020/2020A/2020B tests. :heavy_check_mark: 
6. If 2020A/B prove useful - freedv-gui integration and release for alpha test with end users. :heavy_check_mark: 

# Conclusions

April 2022, this work was written up in this [blog post](https://www.rowetel.com/?p=8014). We'll keep 2020B, but not 2020A.

### CODE REVIEW

**Comment by tmiw** [comment]
@drowe67, should I roll a test build of freedv-gui with 2020A instead of 2020? Or too soon?

**Comment by drowe67** [comment]
> @drowe67, should I roll a test build of freedv-gui with 2020A instead of 2020? Or too soon?

I'm still building 2020A, so there isn't actually a 2020A mode at the FreeDV API level yet...

**Comment by drowe67** [comment]
@srsampson it's getting very easy to work up new OFDM modem waveforms.

**Comment by tmiw** [comment]
> > @drowe67, should I roll a test build of freedv-gui with 2020A instead of 2020? Or too soon?
> 
> I'm still building 2020A, so there isn't actually a 2020A mode at the FreeDV API level yet...

Figured as much. Looks good so far. 👍

**Comment by srsampson** [comment]
> @srsampson it's getting very easy to work up new OFDM modem waveforms.

The one thing that might be beneficial is an easy way to operate with more than 2 bits per symbol!

But very happy with the OFDM!

**Comment by drowe67** [comment]
> > > @drowe67, should I roll a test build of freedv-gui with 2020A instead of 2020? Or too soon?
> > 
> > 
> > I'm still building 2020A, so there isn't actually a 2020A mode at the FreeDV API level yet...
> 
> Figured as much. Looks good so far. +1

I have the clipper (PAPR compression) running for 2020 at the FreeDV API level, which has made quite a difference in 2020 performance.  That  could be tried out at any time with freedv-gui.  You should see the RMS power jump up with clipping enabled, and the scatter diagram get a bit messier without noise.

Try clip 0/1 and -26/-30 for the noise level:
```
cd codec2/build_linux/src
cat ~/LPCNet/wav/wia.wav | ./freedv_tx 2020 - - --clip 0 | ./cohpsk_ch - - -26 --Fs 8000 --fading_dir ../unittest --mpp | ./freedv_rx 2020 - - | aplay -f S16_LE -r 16000
```

**Comment by drowe67** [comment]
Baseline test before starting unequal protection experiments.  Lets try 2020 with and without compression, and compare to a lower rate waveform.

| Test | Mode | compression | No | SNR  | CBER | CPER
| --- | --- | --- | --- | --- | --- | --- | 
| 1 | 2020 | 0 | -32 | 12.3 | 0.01 | 0.20 |
| 2 | 2020 | 1 |  -28 | 13.0 | 0.01| 0.18 |
| 3 | 700E | 0 |  -26 | 6.9 | 0.01 | 0.09 |
| 4 | 700E | 1 |  -22 | 7.8 | 0.01 | 0.10 |

Notes:
1. All tests have the same peak power, which models a real world SSB transmitter.  The noise `No` is varied until we hit coded BER=0.01, or coded PER=0.1.  This threshold has been established via experimentation with the low rate voice modes.
1. 700E is used as the "control", although arguably 700D could have been used, as it's waveform design (e.g. pilot symbol rate) is closer to 2020.
1. Typical command line:
   ```
   ./src/freedv_tx 700E ../raw/ve9qrp.raw - --testframes --clip 1 | ./src/ch - - --No -25 --mpp | ./src/freedv_rx 700E - /dev/null --testframe
   ```
1. The Multipath Poor (MPP) channel was used for these tests,  We can see than vanilla uncompressed 2020) is about 10dB behind 700E in terms of the `No` level for the channel.  This might explain the lack of on-air robustness for FreeDV 2020.
1. Even at a low No/high SNR, the CPER doesn't get down to the target 0.1 - we don't have a very strong FEC code and it's hard to protect this mode against multipath due to the high bit rate required.
1. Adding compression to 2020 results in a useful 4dB gain, similar to 700E.

**Comment by drowe67** [comment]
I've made a first pass at an [unequal error protection](https://github.com/drowe67/LPCNet/pull/42#issuecomment-1003471575) scheme applied to 2020.  This applies very strong FEC to the 1st stage vector index. 

Initial results are encouraging, 2020 is producing distorted but usable speech at the same SNR as 700E (SNR=7dB, No = -22) on the MPP channel.

Here are plots of the output speech for the equal (vanilla 2020) and unequal protection schemes.  The strong protection on the first VQ index results in less gross amplitude errors, and the decoded speech sounds better than with error protection applied equally to all bits.  Both samples used compression (clipping) of the modem waveform to enhance the average power.

Screenshot from 2022-01-13 06-55-27

| Test | Mode | Comp | No | SNR | FEC | Comment |
| --- | --- | --- | --- | --- | --- | --- |
| [1](https://rowetel.com/downloads/deep/2020A/peter_highsnr.wav)  | 2020 | 0 | -100 | 80 | equal | vanilla 2020 with no compression (what we run today) |
| [2](https://rowetel.com/downloads/deep/2020A/peter_equal_noclip.wav) | 2020 | 0 | -22 | 2.4 | equal | vanilla 2020 at 700E operating point |
| [3](https://rowetel.com/downloads/deep/2020A/peter_equal.wav) | 2020 | 1 | -22 | 6.9 | equal | 2020 with compression |
| [4](https://rowetel.com/downloads/deep/2020A/peter_unequal.wav) | 2020 | 1 | -22 | 6.9 | unequal | 2020 with compression and unequal error protection |
| [5](https://rowetel.com/downloads/deep/2020A/peter_700E.wav) | 2020 | 1 | -22 | 8.0 | equal | 700E at same operating point SNR |

**Comment by drowe67** [comment]
@tmiw - just a heads up that I'm building up some experimental 2020-ish modes in this PR.  There are several ideas I'm trying out (compression, VQ index optimisation, unequal error protection, and a waveform that can handle faster fading).

It's unclear which (if any) will actually lead to real improvements at this stage.  Once I've determined what works (and what doesn't) - we can decide which bits of this PR make sense to merge.  Such is R&D :slightly_smiling_face:

**Comment by drowe67** [comment]
@tmiw - I plan to have 2020A/B working at the FreeDV API level soon.  If you feel like working on it, this could be integrated with freedv-gui as a (very) experimental release.  

We would need to make it clear there is no guarantee 2020A/B will be retained, and they might be subject to change at any time.  I would also be wary of running them with multi-rx, as the unique words are the same, and the CPU load is pretty high when the LPCNet decoder kicks in which complicates life.

2020A/B might be fun to play with on the activity day. If there is a clear advantage, it will show up in adhoc OTA tests. I'll also be doing some controlled tests using automated methods (KiwiSDRs etc).

Here are three main innovations, and the theoretical improvements:
1. Compression (clipping) of the 2020x modem waveforms has been added, which is worth about 4dB.  Should improve vanilla 2020 straight away.
1. 2020A uses the same waveform as 2020, but an unequal error protection scheme.  The most important 11 bits of the 52 bit LPCNet codec payload are heavily protected, the other bits not protected at all.  This changes the "slope" of the speech quality against SNR curve.  Compared to 2020, it may work (with poor speech quality) at lower SNRs, but still have a few audible errors even at higher SNRs.
2. 2020B is like 700E to 700D - it works with fast fading but requires a few more dB.  This will make it usable in European Winter (or over the South Pole Argentina to Australia) type channels - if you have enough SNR.  The big challenge here was squeezing all the info we need (enough pilots symbols for fast fading, LPCNet, FEC bits) into a 2100 Hz channel - we are pushing up again the edges of many SSB filters. It also uses unequal FEC, just the most important 11 bits are protected, but not as well as 2020A.

**Comment by drowe67** [comment]
Testing 2020B on a MPD channel (2Hz fading, 4ms delay) at 10dB SNR, compressed modem waveform 5dB PAPR:

`cat ~/LPCNet/wav/all.wav | ~/LPCNet/build_linux/src/lpcnet_enc -x | ./src/ofdm_mod --mode 2020B --ldpc --clip --txbpf | ./src/ch - - --No -25 --mpd | ./src/ofdm_demod --mode 2020B --verbose 1 --ldpc | ~/LPCNet/build_linux/src/lpcnet_dec -x | aplay -f S16_LE -r 16000`

Audio is rough but intelligible.  No loss of sync. We haven't build 2020B into the FreeDV API yet, so quite a few command line steps to stick it all together.

**Comment by drowe67** [comment]
FreeDV 2020B demo on MPD channel at 10dB SNR:

```
./src/freedv_tx 2020B ~/LPCNet/wav/peter.wav - --clip 1  | ./src/ch - - --No -25 --mpd | ./src/freedv_rx 2020B - - | aplay -f S16_LE -r 16000
```

**Comment by drowe67** [comment]
@tmiw - In https://github.com/drowe67/codec2/pull/274/commits/05d3423d4df439e05a94933247f3e2258dc1d2a1 we have a first pass of 2020A & 2020B running at the FreeDV API level. I'll refine it over the few days, but that wont change the API much, so I figure it's ready to try with freedv-gui if you have the time/interest :slightly_smiling_face:

**Comment by tmiw** [comment]
> @tmiw - In [05d3423](https://github.com/drowe67/codec2/commit/05d3423d4df439e05a94933247f3e2258dc1d2a1) we have a first pass of 2020A & 2020B running at the FreeDV API level. I'll refine it over the few days, but that wont change the API much, so I figure it's ready to try with freedv-gui if you have the time/interest 🙂

Hmm, doesn't look like there's a 2020B #define in this branch?

```
[ 47%] Building CXX object src/CMakeFiles/FreeDV.dir/main.cpp.o
/Users/mooneer/devel/freedv-gui/src/main.cpp:917:22: error: use of undeclared identifier 'FREEDV_MODE_2020B'
                case FREEDV_MODE_2020B:
                     ^
/Users/mooneer/devel/freedv-gui/src/main.cpp:931:37: error: use of undeclared identifier 'FREEDV_MODE_2020B'
                    (currentMode == FREEDV_MODE_2020B)) {
                                    ^
/Users/mooneer/devel/freedv-gui/src/main.cpp:1473:18: error: use of undeclared identifier 'FREEDV_MODE_2020B'
        g_mode = FREEDV_MODE_2020B;
                 ^
/Users/mooneer/devel/freedv-gui/src/main.cpp:1537:54: error: use of undeclared identifier 'FREEDV_MODE_2020B'
            g_mode == FREEDV_MODE_2020A || g_mode == FREEDV_MODE_2020B || /* note: 2020A/B don't play well with multi-RX atm */
                                                     ^
4 errors generated.
```

**Comment by drowe67** [comment]
Hmm, I can see the #define [here](https://github.com/drowe67/codec2/pull/274/files#diff-ce88e396a38ebe9339d926fdb4db19df880c5c25467ffb7bb8cae44728dd39ca).

I just checked it out a fresh copy and it's building OK for me...

**Comment by tmiw** [comment]
I tried again and it built okay. Next problem:

```
Assertion failed: ((vq_type == LPCNET_PRED) || (vq_type == LPCNET_DIRECT_SPLIT) || (vq_type == LPCNET_DIRECT_SPLIT_INDEX_OPT)), function lpcnet_quant_create, file lpcnet_quant.c, line 35.
Abort trap: 6
```

Partial backtrace:

```
0   libsystem_kernel.dylib        	    0x7ff81e032112 __pthread_kill + 10
1   libsystem_pthread.dylib       	    0x7ff81e068214 pthread_kill + 263
2   libsystem_c.dylib             	    0x7ff81dfb4d10 abort + 123
3   libsystem_c.dylib             	    0x7ff81dfb40be __assert_rtn + 314
4   liblpcnetfreedv.0.2.dylib     	       0x10af08a01 lpcnet_quant_create.cold.1 + 33 (lpcnet_quant.c:34)
5   liblpcnetfreedv.0.2.dylib     	       0x10aefd6cb lpcnet_quant_create + 523 (lpcnet_quant.c:34)
6   liblpcnetfreedv.0.2.dylib     	       0x10af02c8c lpcnet_freedv_create + 44 (lpcnet_freedv.c:22)
7   libcodec2.1.0.dylib           	       0x108fd679d freedv_2020x_open + 909 (freedv_2020.c:131)
8   libcodec2.1.0.dylib           	       0x108fd0aa8 freedv_open_advanced + 104 (freedv_api.c:163)
9   FreeDV                        	       0x10719aa02 FreeDVInterface::start(int, int, bool, bool) + 402 (freedv_interface.cpp:128)
10  FreeDV                        	       0x10711a6e1 MainFrame::OnTogBtnOnOff(wxCommandEvent&) + 1457 (main.cpp:1584)
```

**Comment by drowe67** [comment]
Not sure.  The asserts says libcodec2 is giving the LPCNet init function the wrong argument at [start up](https://github.com/drowe67/codec2/pull/274/files#diff-23f122f9e93895a0380c373fd9c97126a3f080e764e05b056c7cb146fb056f70R131).  .... maybe another build problem or an earlier libcodec2 is being used?

Does `freedv_rx/rx` work OK from the command line, e.g.:
```
./src/freedv_tx 2020B ~/LPCNet/wav/peter.wav - --clip 1 | ./src/ch - - --No -25 --mpd | ./src/freedv_rx 2020B - - | aplay -f S16_LE -r 16000
```

**Comment by tmiw** [comment]
Hmm, it looks like the test apps behave properly:

```
HT-TM05:build_osx mooneer$ sox ../../LPCNet/wav/all.wav -b 16 -e signed-integer --endian little  all.raw
HT-TM05:build_osx mooneer$ src/freedv_tx 2020A all.raw all_tx.raw
HT-TM05:build_osx mooneer$ src/freedv_rx 2020A all_tx.raw all_rx.raw
frames decoded: 276  output speech samples: 794880
HT-TM05:build_osx mooneer$ src/freedv_tx 2020B all.raw all_tx.raw
HT-TM05:build_osx mooneer$ src/freedv_rx 2020B all_tx.raw all_rx.raw
frames decoded: 552  output speech samples: 794880
HT-TM05:build_osx mooneer$
```

I also took a closer look at the PR, specifically in freedv_tx.c and freedv_rx.c. It seems that `adv.lpcnet_vq_type` is being set prior to `freedv_open_advanced()`, so perhaps freedv-gui just needs to set it to 1 (the default for 2020A/B per the help output).

**Comment by tmiw** [comment]
Yeah, setting `adv.lpcnet_vq_type` seems to have resolved the crashes. 👍

**Comment by drowe67** [comment]
> I also took a closer look at the PR, specifically in freedv_tx.c and freedv_rx.c. It seems that `adv.lpcnet_vq_type` is being set prior to `freedv_open_advanced()`, so perhaps freedv-gui just needs to set it to 1 (the default for 2020A/B per the help output).

Actually the default for 2020 A and B is "index optimization on", which is a **2** in the [first field](https://github.com/drowe67/codec2/pull/274/files#diff-0e1ab0187321cc1b6f05dc99ed54ba42d403b302124c90d26aad37e4931dbadeR121) of `freedv_advanced adv`.  The #defines are spelt out properly [here](https://github.com/drowe67/LPCNet/pull/42/files#diff-17917e96232aa697f318461704737980fa1974317a71781fc3f0b351f6ca31e6R16).  Sorry I should have used the #defines in `freedv_api.c`

These defaults are used if you call `freedv_open()` (non advanced version).  I only use  `freedv_open_advanced()` to selectively enable/disable index optimisation selection, as I would like to try both versions for my current test campaign.  Suggest you just use the default index optimisation by calling `freedv_open()` for `freedv-gui`. It's all a bit confusing atm, but we can refactor this to make it clearer if we find the new 2020 modes are worth keeping.

**Comment by tmiw** [comment]
> > I also took a closer look at the PR, specifically in freedv_tx.c and freedv_rx.c. It seems that `adv.lpcnet_vq_type` is being set prior to `freedv_open_advanced()`, so perhaps freedv-gui just needs to set it to 1 (the default for 2020A/B per the help output).
> 
> Actually the default for 2020 A and B is "index optimization on", which is a **2** in the [first field](https://github.com/drowe67/codec2/pull/274/files#diff-0e1ab0187321cc1b6f05dc99ed54ba42d403b302124c90d26aad37e4931dbadeR121) of `freedv_advanced adv`. The #defines are spelt out properly [here](https://github.com/drowe67/LPCNet/pull/42/files#diff-17917e96232aa697f318461704737980fa1974317a71781fc3f0b351f6ca31e6R16). Sorry I should have used the #defines in `freedv_api.c`
> 
> These defaults are used if you call `freedv_open()` (non advanced version). I only use `freedv_open_advanced()` to selectively enable/disable index optimisation selection, as I would like to try both versions for my current test campaign. Suggest you just use the default index optimisation by calling `freedv_open()` for `freedv-gui`. It's all a bit confusing atm, but we can refactor this to make it clearer if we find the new 2020 modes are worth keeping.

Agreed and updated the other PR accordingly.

**Comment by tmiw** [comment]
Removed the UI for 2020A over at https://github.com/drowe67/freedv-gui/pull/211.

File: README_freedv.md [hunk 1]
```diff
@@ -54,11 +54,13 @@ These are designed for use with a HF SSB radio.
...
-| 700C | 2017 | Codec2 700C | 14 carrier coherent QPSK + diversity | 1500 | 1400 | - | - | 2 | good |
-| 700D | 2018 | Codec2 700C | 17 carrier coherent OFDM/QPSK | 1000 | 1900 | LDPC (224,112)  | 25 | -2 | fair |
-| 700E | 2020 | Codec2 700C | 21 carrier coherent OFDM/QPSK | 1500 | 3000 | LDPC (112,56)  | 25 | 1 | good |
-| 2020 | 2019 | LPCNet 1733 | 31 carrier coherent OFDM/QPSK | 1600 | 3000 | LDPC (504,396)  | 22.2 | 2 | poor |
+| 1600  | 2012 | Codec2 1300 | 14 DQPSK + 1 DBPSK pilot carrier | 1125 | 1600 | Golay (23,12) | 25 | 4 | poor |
+| 700C  | 2017 | Codec2 700C | 14 carrier coherent QPSK + diversity | 1500 | 1400 | - | - | 2 | good |
+| 700D  | 2018 | Codec2 700C | 17 carrier coherent OFDM/QPSK | 1000 | 1900 | LDPC (224,112)  | 25 | -2 | fair |
+| 700E  | 2020 | Codec2 700C | 21 carrier coherent OFDM/QPSK | 1500 | 3000 | LDPC (112,56)  | 25 | 1 | good |
+| 2020  | 2019 | LPCNet 1733 | 31 carrier coherent OFDM/QPSK | 1600 | 3000 | LDPC (504,396)  | 22.2 | 2 | poor |
+| 2020A | 2022 | LPCNet 1733 | 31 carrier coherent OFDM/QPSK | 1600 | 3000 | LDPC (504,396) unequal  | 22.2 | 2 | fair |
```
Comment (tmiw) [inline README_freedv.md:62]: Remove 2020A?

**Pull request merged**